server/route.go

```diff
@@ -33,12 +117,209 @@ func (r *Route) equals(other *Route) bool {
+			logger.Debug("Closing route cause of full queue")
+			r.Close()
+			mTotalDeliverMessageErrors.Add(1)
+			return ErrQueueFull
+		}
+	}
+
+	r.queue.push(m)
+	logger.Debug("Queue size: %d", r.queue.size())
+
+	if !r.isConsuming() {
+		logger.Debug("Starting consuming")
+		go r.consume()
+	}
+
+	runtime.Gosched()
+	return nil
+}
+
+// consume starts to consume the queue and pass the messages to route
+// channel. Stops if there are no items in the queue. Should be started as a goroutine.
+func (r *Route) consume() {
+	r.logger.Debug("Consuming route queue")
+
+	r.setConsuming(true)
+	defer r.setConsuming(false)
+
+	var (
+		m   *protocol.Message
+		err error
+	)
+
+	for {
+		if r.isInvalid() {
+			r.logger.Debug("Stopping consuming cause route is invalid.")
+			mTotalDeliverMessageErrors.Add(1)
+			return
+		}
+
+		m, err = r.queue.poll()
+
+		if err != nil {
+			if err == errEmptyQueue {
+				r.logger.Debug("Empty queue")
+				return
+			}
+			r.logger.WithField("error", err).Error("Error fetching queue message")
+			continue
+		}
+
+		// send next message throught the channel
+		if err = r.send(m); err != nil {
+			r.logger.WithField("message", m).Error("Error sending message through route")
+			if err == errTimeout || err == ErrInvalidRoute {
+				// channel been closed, ending the consumer
+				return
+			}
+		}
+		// all good shift the first item out of the queue
+		r.queue.remove()
+	}
+}
+
+func (r *Route) send(m *protocol.Message) error {
+	defer r.invalidRecover()
+	r.logger.WithField("message", m).Debug("Sending message through route channel")
+
+	// no timeout, means we dont close the channel
+	if r.timeout == -1 {
+		r.messagesC <- r.format(m)
+		r.logger.WithField("size", len(r.messagesC)).Debug("Channel size")
+		return nil
+	}
+
+	select {
+	case r.messagesC <- r.format(m):
+		return nil
+	case <-r.closeC:
+		return ErrInvalidRoute
+	case <-time.After(r.timeout):
+		r.logger.Debug("Closing route cause of timeout")
+		r.Close()
+		return errTimeout
+	}
+}
+
+// recover function to use when sending on closed channel
+func (r *Route) invalidRecover() error {
+	if rc := recover(); rc != nil && r.isInvalid() {
+		r.logger.WithField("error", rc).Debug("Recovered closed router")
+		return ErrInvalidRoute
+	}
+	return nil
+}
+
+// format message according to the channel format
+// TODO: this will be dropped cause it's used only in GCM and after implementing
+// separated subscriptions wont be required anymore
+func (r *Route) format(m *protocol.Message) *MessageForRoute {
+	return &MessageForRoute{Message: m, Route: r}
+}
+
+// sendDirect sends the message directly in the channel
+func (r *Route) sendDirect(m *protocol.Message) error {
+	select {
+	case r.messagesC <- r.format(m):
+		return nil
+	default:
+		r.logger.Debug("Closing route cause of full channel")
+		r.Close()
+		return ErrChannelFull
+	}
+}
+
 // Close closes the route channel.
-func (r *Route) Close() {
+func (r *Route) Close() error {
```

Comment:
suggestion: you could move the exported functions before the private ones, inside the files